scripts/generate_object_properties.cr

```diff
@@ -0,0 +1,862 @@
+      #   end
+      #
+      #   def #{@method_prefix}name?
+      #     #{@var_prefix}name
+      #   end
+      #
+      #   def #{@method_prefix}name
+      #     #{@var_prefix}name.not_nil!
+      #   end
+      # end
+      # ```
+      macro #{@macro_prefix}property!(*names)
+        {% for name in names %}
+    #{def_vars!}
+    #{def_getter!}
+    #{def_setter}
+        {% end %}
+      end
+    TEXT
+  end
+
+  def gen_setter
+    puts <<-TEXT
+      # Defines setter methods for each of the given arguments.
+      #
+      # Writing:
+      #
+      # ```
+      # class Person
+      #   #{@macro_prefix}setter name
+      # end
+      # ```
+      #
+      # Is the same as writing:
+      #
+      # ```
+      # class Person
+      #   def #{@method_prefix}name=(#{@var_prefix}name)
+      #   end
+      # end
+      # ```
+      #
+      # The arguments can be string literals, symbol literals or plain names:
+      #
+      # ```
+      # class Person
+      #   #{@macro_prefix}setter :name, "age"
+      # end
+      # ```
+      #
+      # If a type declaration is given, a variable with that name
+      # is declared with that type.
+      #
+      # ```
+      # class Person
+      #   #{@macro_prefix}setter name : String
+      # end
+      # ```
+      #
+      # is the same as writing:
+      #
+      # ```
+      # class Person
+      #   #{@var_prefix}name : String
+      #
+      #   def #{@method_prefix}name=(#{@var_prefix}name : String)
+      #   end
+      # end
+      # ```
+      #
+      # The type declaration can also include an initial value:
+      #
+      # ```
+      # class Person
+      #   #{@macro_prefix}setter name : String = "John Doe"
+      # end
+      # ```
+      #
+      # Is the same as writing:
+      #
+      # ```
+      # class Person
+      #   #{@var_prefix}name : String = "John Doe"
+      #
+      #   def #{@method_prefix}name=(#{@var_prefix}name : String)
+      #   end
+      # end
+      # ```
+      #
+      # An assignment can be passed too, but in this case the type of the
+      # variable must be easily inferable from the initial value:
+      #
+      # ```
+      # class Person
+      #   #{@macro_prefix}setter name = "John Doe"
+      # end
+      # ```
+      #
+      # Is the same as writing:
+      #
+      # ```
+      # class Person
+      #   #{@var_prefix}name = "John Doe"
+      #
+      #   def #{@method_prefix}name=(#{@var_prefix}name)
+      #   end
+      # end
+      # ```
+      macro #{@macro_prefix}setter(*names)
+        {% for name in names %}
+          {% if name.is_a?(TypeDeclaration) %}
+            {% var_name = name.var.id %}
+            {% type = name.type %}
+            #{@var_prefix}{{name}}
+          {% elsif name.is_a?(Assign) %}
+            {% var_name = name.target %}
+            {% type = nil %}
+            #{@var_prefix}{{name}}
+          {% else %}
+            {% var_name = name.id %}
+            {% type = nil %}
+          {% end %}
```

Comment:
The above is duplicating `def_vars` with the sole difference being that we don't check if the `block` macro variable is defined because there are none for setter. We can avoid the duplication.